### ISSUE TITLE
Add authentication/authorization to the API

### ISSUE THREAD

**Issue by mikesmit** [PolicyEngine/policyengine-api#2063]
Currently we do not expect or process any authentication token from the clients using our API. The scope of this task is to update the API to process the authentication token, when provided, and then use it to authorize a subset of operations where appropriate.

Scope includes determining which operations should be constrained with authorization, but that must include at least editing a user profile.

**Comment by mikesmit** [comment]
picking up the first iteration of this. Looking first at what we do with the household API.

**Comment by mikesmit** [comment]
## TLDR
I have a branch we can use to test auth. Currently the back end will not accept our tokens.

Probable next steps
1. Fix [typo setting header to "Authentication" instead of "Authorization"](https://github.com/PolicyEngine/policyengine-app/blob/ab938945b9ccbcef23031b9616f9478e0b4642e0/src/hooks/useAuthenticatedFetch.js#L21) (does not block testing. see below)
2. Follow  [this guide](https://auth0.com/docs/get-started/architecture-scenarios/spa-api/part-2) to define and configure the API for the SPA in auth0, but there are problems
  * It explicitly recommends a flow for SPA that (When you click on the link) explicitly tells you not to use that flow for SPA.
  * I don't think we want/need all the stuff it makes you set up, but the auth libraries seem to want it (unclear if it's impossible to avoid or just not clearly documented)
  * I don't know what it costs to turn on the authorization extension (required by the above)

## The current error
Ignoring the header type issue (which I fixed by fixing my curl command).

The immediate error is that the token has a header "alg" which is set to "dir" and the parser library for the token does not support that option. (I have directly confirmed this error via debug)

**why is it set to 'dir'?**
Based on this post [here](https://community.auth0.com/t/jwt-alg-field-mismatch-issue-with-auth0-and-python-backend/107182) (I have not independently confirmed this is true) this is what the client library does unless you specify an "audience" (the API).

At least according to the [SPA API Guide](https://auth0.com/docs/get-started/architecture-scenarios/spa-api), doing that requires enabling the "Authorization extension" which we don't currently have and which provides a lot more functionality than we actually need or want.


## What I have and how to try it

0. Download [branch 2063_add_auth_to_api](https://github.com/PolicyEngine/policyengine-api/tree/2063_add_auth_to_api) of the API.
1. Go to our website (either the public link or a local running instance)
2. navigate to the user profile with the debug window open
3. in the "network" tab of the browser debug find the request for the user-profile resource
4. right click and copy as "curl"
5. edit the curl as follows
   6. remove all the headers except Authentication
   7. rename "Authentication" to "Authorization"
   8. change the URL to your locally running API (http://localhost:5000/...)
6. Run the curl
  7. to actually find the detailed root cause for errors I had to run in debug mode and trace errors. In vscode this launch worked for me:
```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python Debugger: Flask",
            "type": "debugpy",
            "request": "launch",
            "module": "flask",
            "env": {
                "FLASK_APP": "policyengine_api.api",
                "FLASK_DEBUG": "1"
            },
            "args": [
                "run",
                "--no-debugger",
                "--no-reload",
                "--without-threads"
            ],
            "jinja": true,
            "autoStartBrowser": false,
            "justMyCode": false
        }
    ]
}
```

**Comment by mikesmit** [comment]
I don't see any updates so I'm picking this back up. 
1. Will put in fix for the UI typo
2. Will confirm the preferred path for enabling SPA auth 
3. Will check the cost of that preferred path and confirm with Max before turning it on.

That should make it possible to test the API with the new token and make any further adjustments required.

**Comment by mikesmit** [comment]
Based on the following, we are using the PKCE login workflow already so no need to change anything there.
1. [Call Your API Using the Authorization Code Flow with PKCE](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-pkce/call-your-api-using-the-authorization-code-flow-with-pkce) points to
2. [Add Login Using the Authorization Code Flow with PKCE](https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-pkce/add-login-using-the-authorization-code-flow-with-pkce) which then says the SDK Does it and points to
3. [Add Login to Your React Application](https://auth0.com/docs/quickstart/spa/react/interactive) 
4. Watched the network traffic when logging in and confirmed the auth CDK redirect includes the code_challenge query parameter.

**Comment by mikesmit** [comment]
Took a stab at setting things up as per the [SPI API documentation](https://auth0.com/docs/get-started/architecture-scenarios/spa-api/part-2).

The first step is to add a new API. It looks like that exceeds our current limit. We may be able to re-use the application we already have at least to demonstrate everything working together.

I suspect we will run into similar issues on the second step: Enabling the Authorization Extension.

**Comment by mikesmit** [comment]
Attempted to install the extension and just got "Error!There was an error submitting the form. Please try again." based on [this discussion](https://community.auth0.com/t/getting-an-error-when-trying-to-install-auth0-authorization-extension/99608/4) that is likely an issue of reaching account limits.

**Comment by mikesmit** [comment]
TLDR: there are two issues with setting up [SPA/API auth](https://auth0.com/docs/get-started/architecture-scenarios/spa-api/part-2) related to resource constraints and we should probably upgrade to the paid version.

Details:
1. **I can't add another API in the Auth0 console due to limits. ** - when you request the JWT for the API you need to provide an "audience" which in this context is an API defined in auth0. We currently define one, household API, and cannot define another under the current constraints. There are two possible paths here
   * **we could use one API for everything** there is not actually any functional reason we _have_ to have two APIs in oauth. We could use household as an audience for both.
   * **upgrade auth0 account** then we could define a second one.
2. **I can't install the "Authorization Extension"** - this is required to associate an SPA application defined in the auth0 console with an API. Currently it fails with a very generic error that discussion on the forum indicates is related to a resource constraint. There are some options here:
   * **clean up some resources** - it's not clear which resources are the issue, but the forum discussion said removing some applications helped.
   * **upgrade auth0 account** presumably this would not be an issue.

**Comment by mikesmit** [comment]
Updated the auth account: 
* Upgraded the auth0 account.
* Enabled the Auth0 Authorization Extension
* Added a new API for the web app api ("https://api.policyengine.org/")
* I **did not** actually configure any App/API permissions, but apparently that's enough.

locally updated the app
* Updated the provider to specify the API as audience
* Updated the fetch to also specify the API as audience
* If I force a re-login, the app will now generate a proper JWT.


Two outcomes
1. I don't actually think we need the Auth0 extension. So it's possible we could do this without upgrading (we'd still need an API)
2. This won't work until all our users get a new refresh token with that audience (forcing a login makes this happen at minimum)

**Comment by mikesmit** [comment]
Verified the new bearer tokens from the front end (PR incoming) work with the back end.

**Comment by mikesmit** [comment]
As I was experimenting with integrating oauth into the fastapi demo, I saw that within the UI apparently useAuth0 can return isAuthenticated as true but getAccessTokenSilently will still fail because of lack of a refresh token.

This means the user can be logged in, but we can't get a bearer token to authenticate with the API.

I will take a look to see what's going on.

**Comment by mikesmit** [comment]
Ok. looked in local storage to see what my token is and then debugged in browser: 
1. The token is expired. (current token expires at Tue Jan 07 2025 21:03:38 GMT-0800 (Pacific Standard Time))
2. isAuthenticated still returns "true" and I was not logged out of the UIX in any way or required to log back in again when I visited the site.

I am now tracing through the auth0 code to understand why we can be authenticated with an expired token.

**Comment by mikesmit** [comment]
Looks like this is by design (?). When loading the current user, Auth0 doesn't appear to check the expiration, just that a user is in the cache.


#### Details
The "authorized" parameter returned by useAuth0(), is set if there is a cached user. (this is done in the [Auth0Provider](https://github.com/auth0/auth0-react/blob/6429fdd31959548906a2ee1ecc8ec4d3a137ebc3/src/auth0-provider.tsx#L135)
```
(async (): Promise<void> => {
      try {
        let user: User | undefined;
        if (hasAuthParams() && !skipRedirectCallback) {
          const { appState } = await client.handleRedirectCallback();
          user = await client.getUser();
          onRedirectCallback(appState, user);
        } else {
          await client.checkSession();
          user = await client.getUser();
        }
        dispatch({ type: 'INITIALISED', user });
      } catch (error) {
        dispatch({ type: 'ERROR', error: loginError(error) });
      }
    })();
```

Which in turn is used in the [reducer](https://github.com/auth0/auth0-react/blob/main/src/reducer.tsx#L31) to determine if we are authenticated or not

If you follow the ``client.checkSession()`` bit you end up finding it is populated via [Auth0Client._getIdTokenFromCache](https://github.com/auth0/auth0-spa-js/blob/main/src/Auth0Client.ts#L1030) which apparently doesn't check the expiration.

```
  private async _getIdTokenFromCache() {
    const audience = this.options.authorizationParams.audience || 'default';

    const cache = await this.cacheManager.getIdToken(
      new CacheKey({
        clientId: this.options.clientId,
        audience,
        scope: this.scope
      })
    );

    const currentCache = this.userCache.get<IdTokenEntry>(
      CACHE_KEY_ID_TOKEN_SUFFIX
    ) as IdTokenEntry;

    // If the id_token in the cache matches the value we previously cached in memory return the in-memory
    // value so that object comparison will work
    if (cache && cache.id_token === currentCache?.id_token) {
      return currentCache;
    }

    this.userCache.set(CACHE_KEY_ID_TOKEN_SUFFIX, cache);
    return cache;
  }
```

When we use  getTokenSilently() to get the token for authenticating against the API however, _it_  (eventually) uses [Auth0Client._getEntryFromCache](https://github.com/auth0/auth0-spa-js/blob/main/src/Auth0Client.ts#L1055) which appears to actually check the expiration date.
```
private async _getEntryFromCache({
    scope,
    audience,
    clientId
  }: {
    scope: string;
    audience: string;
    clientId: string;
  }): Promise<undefined | GetTokenSilentlyVerboseResponse> {
    const entry = await this.cacheManager.get(
      new CacheKey({
        scope,
        audience,
        clientId
      }),
      60 // get a new token if within 60 seconds of expiring
    );
...
```
```

**Comment by mikesmit** [comment]
Looks like we have enabled refresh tokens in the auth0 account, but just aren't using them. I will put in a PR.

**Comment by nikhilwoodruff** [comment]
Dropping into the current sprint by default and assigning @mikesmit by default if you're able to take this from this context- but please LMK if not, no problem.

**Comment by mikesmit** [comment]
Punding to next sprint again.
